compiler/optimizations/loopInvariantCodeMotion.cpp

```diff
@@ -932,44 +933,48 @@ static void computeLoopInvariants(std::vector<SymExpr*>& loopInvariants,
     // For GPU-bound loops, relax this restriction and say that changing things
     // passed to a kernel by reference from underneath the kernel is undefined
     // behavior.
-    if (isArgSymbol(symExpr->symbol()) &&
-        symExpr->getValType()->symbol->hasFlag(FLAG_ITERATOR_CLASS) == false) {
-      if(ArgSymbol* argSymbol = toArgSymbol(symExpr->symbol())) {
-        if(argSymbol->isRef() && !isGpuBound) {
-          mightHaveBeenDeffedElseWhere = true;
-        }
-      }
-      for_set(Symbol, aliasSym, aliases[symExpr->symbol()]) {
-        if(ArgSymbol* argSymbol = toArgSymbol(aliasSym)) {
-          if(argSymbol->isRef() && !isGpuBound) {
+    auto isArray = symExpr->typeInfo() &&
+                   symExpr->typeInfo()->symbol->hasFlag(FLAG_ARRAY);
```

Comment:
Thanks! This'll only check for the array record. But the idea is by allowing the array record to pass through, we also allow the array instance to pass through as well? IOW, it is not clear to me whether this will hoist both the record and instance, or just the record.

We have `isArrayClass(Type*)` to check whether a type is an array instance type. You may need to consider using that in this check.

Reply:
It won't hoist one without the other, so that's not a concern.

The problem has been that array arguments passed by ref are considered loop variant, and ignored. Since each array access generates its own `.instance` calls, those should occur within the body of the loop given an existing array ref argument; in other words, assuming a loop isn't "capturing" actual instance record variables from outside (which it can only do if a user explicitly writes such a thing, I believe), the check of the array record is _sufficient_. It would also be correct to use `isArrayClass`, though, so I would be happy to switch it.